spinalcordtoolbox/scripts/sct_dmri_moco.py

```diff
@@ -185,7 +185,7 @@ def main(argv=None):
     # Fetch user arguments
     param.fname_data = arguments.i
     param.fname_bvecs = os.path.abspath(arguments.bvec)
-    param.fname_bvals = arguments.bval
+    param.fname_bvals = os.path.abspath(arguments.bval)
```

Comment:
Normally this wouldn't hurt, but in this case if `bval` is not specified by the user it is set to an empty string automatically by `moco.py` (this is not a sane default value): https://github.com/neuropoly/spinalcordtoolbox/blob/67170203934274498b90bc65f6a084f724b6ec19/spinalcordtoolbox/moco.py#L56

If expanded it resolves the absolute path of the current working directory (normally) which is wrong.

Looking at the batch processing log (https://github.com/neuropoly/spinalcordtoolbox/runs/2511298710?check_suite_focus=true#step:4:3136) it seems to have been resolved to `sct_example_data/dmri`.. which I don't understand how, maybe because of the use of `chdir()` in the code? (if this is the case then it's definitely a recipe for disaster)

Reply:
I should have waited for the CI to pass before requesting a review, aha. :sweat_smile: 

> it seems to have been resolved to sct_example_data/dmri.. which I don't understand how, maybe because of the use of chdir() in the code? (if this is the case then it's definitely a recipe for disaster

`batch_processing.sh` does all of its processing directly in `sct_example_data`:

https://github.com/neuropoly/spinalcordtoolbox/blob/67170203934274498b90bc65f6a084f724b6ec19/batch_processing.sh#L186

Reply:
I'm not a fan of the proposed fix. It is necessary only because the code performs unexpected `chdir()` which causes a maintenance and debugging burden.

Reply:
> `batch_processing.sh` does all of its processing directly in `sct_example_data`:

Ah yes I see.

Reply:
> I'm not a fan of the proposed fix. It is necessary only because the code performs unexpected `chdir()` which causes a maintenance and debugging burden.
> 
> edit: not saying that the fix will not work, but if we should at least consider refactoring the `chdir()` in another issue if not in this one.

I had the exact same thought. See the extended commit message in b6ebe82.

Reply:
I wasn't sure how to scope this, as I have a feeling `moco.py` could use a larger refactor. It does a lot of file I/O related tasks that might be better suited for `sct_dmri_moco`. But, I didn't want this to spiral. :p

Reply:
> we should at least consider refactoring the chdir() in another issue if not in this one.

I did a quick regex search for `file_.* =` and `fname_.* =` and there are 30+ instances of relative filepaths being set in `moco_wrapper` -- we'd have to modify all of them if we stopped changing the working directory. It's doable, but it would expand the scope of this PR considerably. And, there are some file paths that are bit tricky to modify... `file` and `fname` don't seem to be used 100% consistently according to the [Naming conventions](https://github.com/neuropoly/spinalcordtoolbox/wiki/Programming%3A-Naming-conventions-(variables,-files)).

I think it might be better to focus on fixing this bug for 5.3.1, and save a refactor for a future PR.

Reply:
> I think it might be better to focus on fixing this bug for 5.3.1, and save a refactor for a future PR.

👍